### PR TITLE
Get system user id in a lazy manner

### PR DESCRIPTION
**Pull request by athos-ribeiro**
Calling getpass.getuser may lead to breakage in environments where there
is no entries in the /etc/passwd file for the current user.

Setting the environment variables for the git user configurations should
prevents GitPython from using values from /etc/passwd. However, doing so
will not prevent reading /etc/passwd and looking for an entry with the
current user UID.

This patch changes the behavior described above so GitPython will
perform a lazy evaluation of /etc/passwd, only doing so when the
environment variables for the git user configuration are not available.

Signed-off-by: Athos Ribeiro <athos@redhat.com>

### CODE REVIEW

**Review by nirzari — APPROVED** [review]

**Comment by Byron** [comment]
Great work, thanks a lot! It's also my first time to see the `nonlocal` keyword, good learning experience.

If this issue is holding you back, I am happy to create a new patch release if you request it.

**Comment by athos-ribeiro** [comment]
> Great work, thanks a lot! It's also my first time to see the `nonlocal` keyword, good learning experience.
> 
> If this issue is holding you back, I am happy to create a new patch release if you request it.

I would really appreciate that!

As soon as that release is out, I will work to push that to Fedora (that is where I consume the package from).

Thank you :)

**Comment by Byron** [comment]
Alright, v3.1.10 was just released to pypi.
Thanks again for your contribution.

**Comment by athos-ribeiro** [comment]
Thanks for that!

Note that a new tag was not pushed to this repository yet.

**Comment by Byron** [comment]
Thanks for the reminder. Done.

**Comment by athos-ribeiro** [comment]
That was fast! I was about to update my comment to say https://gitpython.readthedocs.io/en/stable/changes.html still needs updating. Let me know if you need any help with that. I am building the new Fedora packages right now.

Thanks again for the quick replies on this one!!!

File: git/util.py [hunk 1]
```diff
@@ -582,8 +582,16 @@ def _from_string(cls, string):
...
+        user_id = None  # We use this to avoid multiple calls to getpass.getuser()
+
+        def default_email():
+            nonlocal user_id
+            if not user_id:
+                user_id = get_user_id()
+            return user_id
+
+        def default_name():
+            default_email().split('@')[0]
```
Comment (harupy) [inline git/util.py:594]: Is `return` missing here?

Comment (harupy) [reply git/util.py:594]: I was investigating an error in our CI and found this patch which is included `gitpython=3.10.0`

https://github.com/mlflow/mlflow/pull/3584/checks?check_run_id=1297036420#step:4:1677

Comment (athos-ribeiro) [reply git/util.py:594]: Guess I messed up on this one :)

Will provide a follow-up fix right away.

I am sorry for this :(

Comment (harupy) [reply git/util.py:594]: no problem :) Thanks for the quick reponse!

Comment (Byron) [reply git/util.py:594]: Thanks everyone. With the next PR to fix this one you will get a new patch release right away.

Comment (athos-ribeiro) [reply git/util.py:594]: Fixing it in https://github.com/gitpython-developers/GitPython/pull/1073.

Once again, sorry for the inconvenient

**Pull request merged**